### ISSUE TITLE
"FilterServletResponseWrapper.flushBuffer()" does not flush the response with empty body

### ISSUE THREAD

**Issue by dabing009** [javamelody/javamelody#836]
We use javamelody with tomcat on our product environment. We configure javamelody as the first filter for tomcat webapp. Recently we found an issue that tomcat webapp REST API returns 304 without body, but finally user gets 204.

We did some trouble-shooting and found when tomcat webapp call response.flushBuffer(), it ends up with calling javamelody "FilterServletResponseWrapper.flushBuffer()", its code likes below:

```
 @Override
	public void flushBuffer() throws IOException {
		if (writer != null) {
			writer.flush();
		} else if (stream != null) {
			stream.flush();
		}
    }
```

Looks it only flush response when the body(either writer or stream) is not empty, for our case, response without body cannot be flushed from here. Then it reaches jersey library, from there it changes the code to 204 due to empty body. 

We think it would be better to change the above code to below:

```
	@Override
	public void flushBuffer() throws IOException {
		if (writer != null) {
			writer.flush();
		} else if (stream != null) {
			stream.flush();
		} else {
			((HttpServletResponse)this.getResponse()).flushBuffer();
		}
	}
```

If no body found just call the wrapped response flushBuffer() to send out the response status code and headers. With the above changes, our user finally got correct response code 304.

**Comment by evernat** [comment]
fixed by e0c2fe0 and ready for the next release (1.78).
A snapshot build including the fix is available at:
http://javamelody.org/downloads/javamelody-core-1.78.0-SNAPSHOT.jar

**Comment by dabing009** [comment]
Thanks for quick reply!

**Comment by evernat** [comment]
The fix of this issue (e0c2fe0) caused [JENKINS-58419](https://issues.jenkins-ci.org/browse/JENKINS-58419).
It is not known if the problem is now in javamelody since 1.78.0 or in the jenkins CAS plugin 1.4.3.

**Comment by evernat** [comment]
[JENKINS-58419](https://issues.jenkins-ci.org/browse/JENKINS-58419) (since monitoring-plugin 1.78) is fixed by 068775a:
MonitoringFilter should not flush buffer when no content.
